xbmc/Util.cpp

```diff
@@ -1824,36 +1810,47 @@ void CUtil::ScanForExternalSubtitles(const std::string& strMovie, std::vector<st
     || item.IsLiveTV()
     || !item.IsVideo())
     return;
-  
-  vector<std::string> strLookInPaths;
-  
-  std::string strMovieFileName;
-  std::string strMoviePath;
-  
-  URIUtils::Split(strMovie, strMoviePath, strMovieFileName);
-  std::string strMovieFileNameNoExt(URIUtils::ReplaceExtension(strMovieFileName, ""));
-  strLookInPaths.push_back(strMoviePath);
-  
+
+  CLog::Log(LOGDEBUG,"%s: Searching for subtitles...", __FUNCTION__);
+
+  std::string strBasePath;
+  std::string strSubtitle = URIUtils::ReplaceExtension(URIUtils::GetFileName(strMovie), "");
+
+  if (item.HasVideoInfoTag())
+    strBasePath = item.GetVideoInfoTag()->m_basePath;
+
+  if (item.IsOpticalMediaFile())
+    strBasePath = item.GetLocalMetadataPath();
+
   CURL url(strMovie);
-  std::string isoFileNameNoExt;
-  if (url.IsProtocol("bluray"))
-      url = CURL(url.GetHostName());
-  //a path inside an iso
-  if (url.IsProtocol("udf"))
+  if (url.IsProtocol("bluray") || url.IsProtocol("udf"))
   {
-    std::string isoPath = url.GetHostName();
-    URIUtils::RemoveSlashAtEnd(isoPath);
-    CFileItem iso(isoPath, false);
+    strBasePath = URIUtils::GetParentPath(url.GetHostName());
+    strSubtitle = URIUtils::ReplaceExtension(GetTitleFromPath(url.GetHostName()), "");
+  }
+
+  if (strBasePath.empty())
+    strBasePath = URIUtils::GetBasePath(strMovie);
+
+  CFileItemList items;
+  vector<std::string> strLookInPaths;
+
+  int flags = DIR_FLAG_NO_FILE_DIRS | DIR_FLAG_READ_CACHE | DIR_FLAG_NO_FILE_INFO;
+
+  if (!strBasePath.empty())
+    CDirectory::GetDirectory(strBasePath, items, g_advancedSettings.m_subtitlesExtensions, flags);
+
+  const char * common_sub_dirs[] = {"subs", "subtitles", "vobsubs", "sub", "vobsub", "subtitle", NULL};
```

Comment:
'not invented here' I know, but might want to just use vector/std::array and a ranged loop for that